hcl/eval/eval.go

```diff
@@ -111,12 +108,44 @@ func (c *Context) PartialEval(expr hclsyntax.Expression) (hclwrite.Tokens, error
 	return engine.Eval()
 }
 
-// GetExpressionTokens gets the provided expression writable tokens.
-func GetExpressionTokens(hcldoc []byte, expr hclsyntax.Expression) (hclwrite.Tokens, error) {
-	exprRange := expr.Range()
+// TokensForValue returns the tokens for the provided value.
+func TokensForValue(value cty.Value) (hclwrite.Tokens, error) {
+	value, marks := value.Unmark()
+	if value.Type() == customdecode.ExpressionClosureType {
+		closureExpr := value.EncapsulatedValue().(*customdecode.ExpressionClosure)
+		for m := range marks {
+			switch v := m.(type) {
+			case ExpressionStringMark:
+				exprRange := closureExpr.Expression.Range()
+				exprData := []byte(v)[exprRange.Start.Byte:exprRange.End.Byte]
+				return TokensForExpressionBytes(exprData)
+			}
+		}
+
+		return TokensForExpression(closureExpr.Expression)
+	}
+	return hclwrite.TokensForValue(value), nil
+}
+
+// TokensForExpression gets the provided expression writable tokens.
+// Beware: This function requires a valid expr.Range(), which means that
+// expr.Range().Filename must be the real file used to parse this expression
+// and the ranges (start and end) points to correct positions.
+// The expression must be an expression parsed from a real file.
+func TokensForExpression(expr hhcl.Expression) (hclwrite.Tokens, error) {
 	filename := expr.Range().Filename
-	exprBytes := hcldoc[exprRange.Start.Byte:exprRange.End.Byte]
-	tokens, diags := hclsyntax.LexExpression(exprBytes, filename, hhcl.Pos{})
+	filedata, err := ioutil.ReadFile(filename)
```

Comment:
Oh the humanity :rofl:

Reply:
haha but this was already here, just moved from the PartialEval() function.

Reply:
I know :laughing: